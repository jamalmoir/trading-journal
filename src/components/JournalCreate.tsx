import React, { useState } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import Types from 'Types';
import { createJournal } from '../redux/actions/journal';
import { JournalAction } from '../redux/reducers/journal';
import { TextInput } from './TextInput';
import { CurrencyCode, isCurrencyCode } from '../utils/moolah';


interface JournalCreateProps {
  className?: string;
  onCreateJournal: (journal: Types.Journal) => null;
  auth: any,
}

const JournalCreateComponent = (props: JournalCreateProps) => {
  const [journalName, setJournalName]: [string, (val: string) => void] = useState('');
  const [journalKind, setJournalKind]: [Types.JournalKind, (val: Types.JournalKind) => void] = useState<Types.JournalKind>('live');
  const [journalCurrency, setJournalCurrency]: [CurrencyCode, (val: CurrencyCode) => void] = useState<CurrencyCode>('USD');

  const updateJournalName = (val: string) => {
    setJournalName(val);
  }

  const updateJournalKind = (val: 'live' | 'demo' | 'backtest') => {
    setJournalKind(val);
  }

  const updateJournalCurrency = (val: string) => {
    if (isCurrencyCode(val)) {
      setJournalCurrency(val);
    }
  }

  const createJournal = () => {
    let journal: Types.Journal = {
      id: null,
      userId: props.auth.uid,
      kind: journalKind,
      currency: journalCurrency,
      name: journalName,
      created: new Date(),
      modified: new Date(),
      tradeCount: 0,
    }

    setJournalName('');
    setJournalKind('live');
    setJournalCurrency('USD');
    props.onCreateJournal(journal);
  }

  return (
    <div className={ props.className + ' input-group mb-3'}>
      <TextInput type="text"
              className="form-control col-8"
              placeholder="Journal name"
              value={ journalName }
              onChange={ (e) => updateJournalName(e.target.value) }
      />
      <select className="form-control custom-select col-4"
              value={ journalKind }
              onChange={ (e) => updateJournalKind(e.target.value as 'live' | 'demo' | 'backtest') }
      >
        <option value="live">Live</option>
        <option value="demo">Demo</option>
        <option value="backtest">Backtest</option>
      </select>
      <TextInput type="text"
              className="form-control col-8"
              placeholder="Currency"
              value={ journalCurrency }
              onChange={ (e) => updateJournalCurrency(e.target.value) }
      />
      <div className="input-group-append">
        <button className="btn btn-outline-primary" type="button" onClick={ createJournal }>Create</button>
      </div>
    </div>
  )
};

const mapStateToProps = (state: Types.RootState) => {
  return {
    auth: state.firebase.auth,
  }
};

const mapDispatchToProps = (dispatch: Dispatch<JournalAction>) => ({
  // @ts-ignore
  onCreateJournal: (journal: Types.Journal) => dispatch(createJournal(journal)),
});

export const JournalCreate = connect(mapStateToProps, mapDispatchToProps)(JournalCreateComponent);