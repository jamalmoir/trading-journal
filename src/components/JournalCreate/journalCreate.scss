@import '../../_settings.scss';

.journal-create {
	display: flex;
	flex-direction: row;
	justify-content: space-between;
	align-items: baseline;

	.journal-create-button {
		
	}
}