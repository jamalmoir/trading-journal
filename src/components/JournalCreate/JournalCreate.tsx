import * as React from 'react'
import { useState } from 'react'
import { connect } from 'react-redux'
import { Dispatch } from 'redux'
import Types from 'Types'
import { createJournal } from '../../redux/actions/journal'
import { JournalAction, JournalState } from '../../redux/reducers/journal'
import { TextInput } from '../TextInput'
import { setUpControls } from '../../utils/utils'
import './journalCreate.scss'
import { Button } from '../Button'
import { SelectInput } from '../SelectInput'

interface JournalCreateProps {
  className?: string
  onCreateJournal: (journal: Types.Journal) => null
  auth: any
  journal?: JournalState //TODO - this should not be ?
}

const initialControls: Types.InputControls = {
  name: {
    value: '',
    errors: [],
    touched: false,
    validationRules: {
      minLength: 2,
      notEmpty: true,
    },
  },
  kind: {
    value: '',
    errors: [],
    touched: false,
    validationRules: {
      minLength: 2,
      notEmpty: true,
    },
  },
  currency: {
    value: '',
    errors: [],
    touched: false,
    validationRules: {
      notEmpty: true,
      minLength: 3,
      maxLength: 3,
      isCurrencyCode: true,
    },
  },
}

export const JournalCreateComponent = (props: JournalCreateProps) => {
  const [controlsValid, setControlsValid] = useState(false)
  const [controls, setControls] = setUpControls(initialControls)

  const kindChoices = [
    { id: 'live', name: 'Live' },
    { id: 'demo', name: 'Demo' },
    { id: 'backtest', name: 'Backtest' },
  ]

  const updateControls = (kind: string, value: any) => {
    setControlsValid(setControls(kind, value))
  }

  const createJournal = () => {
    let journal: Types.Journal = {
      id: null,
      userId: props.auth.uid,
      kind: controls.kind.value,
      currency: controls.currency.value,
      name: controls.name.value,
      created: new Date(),
      modified: new Date(),
      tradeCount: 0,
    }
    props.onCreateJournal(journal)
    setControls(null, initialControls)
  }

  return (
    <div className={props.className + ' journal-create grid-x grid-margin-x'}>
      <TextInput
        type="text"
        className="journal-create-name cell small medium-3"
        label="Journal name"
        placeholder="My Journal"
        value={controls.name.value || ''}
        onChange={e => updateControls('name', e.target.value)}
        errors={controls.name.errors}
        touched={controls.name.touched}
      />
      <SelectInput
        label="Kind"
        className="journal-create-kind cell small medium-3"
        placeholder="live"
        value={controls.kind.value}
        onChange={e =>
          updateControls('kind', e.target.value as 'live' | 'demo' | 'backtest')
        }
        errors={controls.kind.errors}
        touched={controls.kind.touched}
        choices={kindChoices}
      />
      <TextInput
        type="text"
        className="journal-create-currency cell small medium-3"
        label="Currency"
        placeholder="USD"
        value={controls.currency.value || ''}
        onChange={e => updateControls('currency', e.target.value)}
        errors={controls.currency.errors}
        touched={controls.currency.touched}
      />
      <Button
        text="Create"
        className="journal-create-button cell small medium-2"
        onClick={createJournal}
        loading={props.journal.isRequesting}
        disabled={
          !controlsValid ||
          !controls.name.value ||
          !controls.kind.value ||
          !controls.currency.value
        }
      />
    </div>
  )
}

const mapStateToProps = (state: Types.RootState) => {
  return {
    auth: state.firebase.auth,
    journal: state.journal,
  }
}

const mapDispatchToProps = (dispatch: Dispatch<JournalAction>) => ({
  // @ts-ignore
  onCreateJournal: (journal: Types.Journal) => dispatch(createJournal(journal)),
})

export const JournalCreate = connect(
  mapStateToProps,
  mapDispatchToProps
)(JournalCreateComponent)
