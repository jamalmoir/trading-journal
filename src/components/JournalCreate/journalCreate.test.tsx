import * as React from 'react';
import { JournalCreateComponent } from './JournalCreate';
import { shallow } from 'enzyme';
import Types from 'Types';


describe("<JournalCreate />", () => {
  it('matches snapshot', () => {
		const onCreateJournal = jest.fn();
    const component = shallow(<JournalCreateComponent onCreateJournal={ onCreateJournal } auth={ {} }/>);
    expect(component).toMatchSnapshot();
	});

  it('calls onCreateJournal with the right arguments', () => {
		const onCreateJournal = jest.fn();
		const component = shallow(<JournalCreateComponent onCreateJournal={ onCreateJournal } auth={ {} }/>);
		const name = 'Test Journal';
		const kind = 'demo';
		const currency = 'GBP';
		const expectedJournal: Types.Journal = {
			id: null,
			userId: null,
			created: null,
			modified: null,
			name: name,
			kind:kind,
			currency: currency,
			tradeCount: null,
		}

		component.find('.journal-create-name').simulate('change', {target: {value: name}});
		component.find('.journal-create-kind').simulate('change', {target: {value: kind}});
		component.find('.journal-create-currency').simulate('change', {target: {value: currency}});
		component.find('button').simulate('click');

		expect(JSON.stringify(onCreateJournal.mock.calls[0][0]) === JSON.stringify(expectedJournal))
	});
});