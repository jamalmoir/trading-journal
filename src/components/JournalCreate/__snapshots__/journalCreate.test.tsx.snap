// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JournalCreate /> matches snapshot 1`] = `
<div
  className="undefined input-group mb-3"
>
  <Component
    className="journal-create-name form-control col-8"
    errors={Array []}
    onChange={[Function]}
    placeholder="Journal name"
    touched={false}
    type="text"
    value=""
  />
  <select
    className="journal-create-kind form-control custom-select col-4"
    onChange={[Function]}
    value="long"
  >
    <option
      value="live"
    >
      Live
    </option>
    <option
      value="demo"
    >
      Demo
    </option>
    <option
      value="backtest"
    >
      Backtest
    </option>
  </select>
  <Component
    className="journal-create-currency form-control col-8"
    errors={Array []}
    onChange={[Function]}
    placeholder="Currency"
    touched={false}
    type="text"
    value=""
  />
  <div
    className="input-group-append"
  >
    <button
      className="btn btn-outline-primary"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Create
    </button>
  </div>
</div>
`;
